---
output: 
  pdf_document: 
    highlight: tango
editor_options: 
  chunk_output_type: console
---
```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.width=6, fig.height=4, fig.path='Figs/',
                      echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE)
```

# Population genetics workshop

Instructor: **Matthias Steinrücken**  
Course Assistant: **Maryn Carlson**  

### Welcome 

This excercise is going to expose you to several basic ideas in probability and statistics as well as show you the utility of using R for basic statistical analyses.  We'll do so in the context of a basic population genetic analysis.  

### The scenario

As a biologist, you will learn what are the major patterns that are expected when the data you work with is clean. Using that expertise will save you from the mistake of misinterpreting error-prone data.   In population genetics, there are a number of patterns that we expect to see immediately in our datasets.  In this exercise you will explore one of those major patterns.  Rather than give it away --- let's begin some analysis and see what we find.  In the narative that follows, we'll refine our thinking as we go.  


### Introductory terminology

- Single-nucleotide polymorphism (SNP):  A nucleotide basepair that is *polymorphic* (i.e. it has multiple types or *alleles* in the population)
- Allele:  A particular variant form of DNA  (e.g. A partciular SNP may have the "A-T" allele in one DNA copy and "C-G" in another; We typically define a reference strand of the DNA to read off of, and then denote the alleles according to the reference strand base - so for example, these might be called simply the "A" and "C" alleles.  In many cases we don't care about the precise base, so we might call these simply the $A_1$ and $A_2$ alleles, or the $A$ or $a$ alleles, or the 0 and 1 alleles.) 
- Minor allele: The allele that is more rare in a population 
- Major allele: The allele that is more common in a population 
- Genotype: The set of alleles carried by an individual (E.g. AA, AC, CC; or AA, Aa, aa; or 0, 1, 2)
- Genotyping array: A technology based on hybridization with probes and florensence that allows genotype calls to be made at 100s of thousands of SNPs per individual at an affordable cost.  

### The data-set and basic pre-processing

We will look at Illumina 650Y genotyping array data from the CEPH-Human Genome Diversity
Panel.  This sample is a global-scale sampling of human diversity with 52 populations in total.

The data were first described in Li et al (Science, 2008) and the raw files are available from the
following link:
<http://hagsc.org/hgdp/files.html>.  These data have been used in numerous subsequent publications (e.g Pickrell et al, Genome Research, 2009) and are an important reference set.  A few technical details are that the genotypes were filtered with a GenCall score cutoff of 0.25 (a quality score generated by the basic genotype calling software).  Individuals with a genotype call rate <98.5% were removed, with the logic being that if a sample has many missing genotypes it may due to poor quality of the source DNA, and so none of the genotypes should be trusted.  Beyond this, to prepare the data for the workshop, we have filtered down the indiviudals to a set of 938 unrelated individuals.  (For those who are interested, the data are available as plink-formatted files `H938.bed` 
`H938.fam`, `H938.bim` from this link: <http://bit.ly/1aluTln>).  We have also extracted the basic counts of three possible genotypes.  

The files with these genotype frequencies are your starting points. 

### Note about logistics

We will use some of functions from the `dplyr` and `ggplot2` and `reshape2` libaries so first let's load them:
```{r}
library(dplyr)
library(ggplot2)
library(reshape2)
```
If you get an error message saying there is no package titled `dplyr`,`ggplot2`, or `reshape2` you may need to first run `install.packages("dplyr")`,`install.packages("ggplot2")`, or`install.packages("reshape2")` to install the appropriate package. 

We will not be outputting files - but you may want to set your working directory to the `sandbox` sub-directory in case you want to output some files. 

The `MBL_WorkshopJN.Rmd` file has the R code that you can run.  I provide code for most steps, but some you will need ot devise for yourselves to answer the questions that are part of the workshop narrative. 

### Initial view of the data

Read in the data table:
```{r}
g <- read.table("../data/H938_chr15.geno", header=TRUE)
```
It will be read in as a dataframe in R.

Then use the "head" command to see the beginning of the dataframe:
```{r}
head(g)
```

You should see that there are columns each with distinct names.  
```
CHR        SNP A1 A2 nA1A1 nA1A2 nA2A2
```
- CHR: The chromosome number.  In this case all SNPs are from chromosome 2.
- SNP: The rsid of a SNP is a unique identifier for a SNP and you can use the rsid to look up information about a SNP using online resource such as dbSNP or SNPedia. 
- A1: The minor allele at the SNP.
- A2: The major allele.
- nA1A1 : The number of A1/A1 homozygotes.
- nA1A2 : The number of A1/A2 heterozygotes.
- nA2A2 : The number of A2/A2 homozygotes.

### Calculate the number of observations at each locus

Next compute the total number of observations by summing each of the three possible genotypes.  Here we use the `mutate` function from the `dplyr` library to do the addition and add a new column to the dataframe in one nice step.  (Note: You could also use the `colSums` function from the base R library).   
```{r}
g <- mutate(g, nObs = nA1A1 + nA1A2 + nA2A2)
```
Run `head(g)` and confirm your dataframe `g` has a new column called `nObs`.

Now use the `summary` function to print a simple summary of the distribution:
```{r}
summary(g$nObs)
```


The `ggplot2` library has the ability to make "quick plots" with the command `qplot`. If we pass it a single column it will make a histogram of the data for that column. Let's try it:

```{r}
qplot(nObs, data = g)
```

Our data are from 938 individuals. When the counts are less than this total, it's because some individuals had array data that was difficult to call a genotype for and so no genotype was reported.  

**Question:**  Do most of the SNPs have complete data?  
**Question:** What is the lowest count observed?   Is this number in rough agreement with what we know about how the genome-wide missingness rate filter was set to >98.5% of all SNPs?

### Calculating genotype and allele frequencies

Let's move on to calculating genotype and allele frequencies.  For allele $A_1$ we will denote its frequency among all the samples as $p_1$, and likewise for $A_2$ we will use $p_2$. 
```{r}
# Compute genotype frequencies
g <- mutate(g, p11 = nA1A1/nObs , p12 = nA1A2/nObs, p22 = nA2A2/nObs )
# Compute allele frequencies from genotype frequencies
g <- mutate(g, p1 = p11 + 0.5*p12, p2 = p22 + 0.5*p12)
```

**Question:**  With a partner or group member, discuss whether the equations in the code for $p_1$ and $p_2$ are correct and if so, why? 

Run `head(g)` again and confirm that `g` now has the extra columns for the genotype and allele frequencies. 

And let's plot the frequency of the major allele (A2) vs the frequency of the minor allele (A1).  The `ggplot2` library has the ability to make "quick plots" with the command `qplot`.  Let's try it here: 

```{r}
qplot(p1, p2, data=g)
```

Notice that $p_2>p_1$ (be careful to inspect the axes labels here)  This makes sense because $A_1$ is supposed to be the minor (less frequent) allele.  Note also that there is a linear relationship between $p_2$ and $p_1$  

**Question:**  What is the equation decribing this relationship?

The relationship exists because there are only two alleles - and so their proportions must sum to 1.  The linear relationship you found exists because of this constraint. It also provides a nice check on our work (if $p_1$ and $p_2$ didn't sum to 1 it would suggest something is wrong with our code!).  

### Plotting genotype on allele frequencies

Let's look at an initial plot of genotype vs allele frequencies.  We could use the base plotting functions, but the following uses the `ggplot2` commands.  These are a little trickier, but end up being very compact (we need fewer lines of code overall to achieve our desired plot). To use `ggplot2` commands effectively our data need to be what statisticans call "tidy" (in this case, that means with one row per pair of points we will plot).  

To do this, we first subset the data on the columns we'd like (using the `select` command and listing the set of columns we want), then we pass this (using the `%>%` operator) to the `melt` command which will reformat the data for us, and output it as `gTidy`:

```{r}
gTidy <- select(g, c(p1,p11,p12,p22)) %>% melt(id='p1',value.name="Genotype.Proportion")
head(gTidy)
ggplot(gTidy) + geom_point(aes(x = p1,
                               y = Genotype.Proportion, 
                               color = variable, 
                               shape = variable))
```

Now let's look at the graph that we produced.  There is some scatter in the relationship between genotype proportion and allele frequency for any given genotype, but at the same time there is a very regular underlying relationship between these variables. 

**Question:** What are approximate relationships between $p_{11}$ vs $p_1$, $p_{12}$ vs $p_1$, and $p_{22}$ vs $p_1$?  (Hint: These look like parabolas, which suggests are some very simple quadratic functions of $p_1$). 

You might start to recognize that these are the classic relationships that are taught in  introductory biology courses.  If you recall, under assumptions that there is no mutation, no natural selection, infinite population size, no population substructure and no migration, then the genotype frequencies will take on a simple relationship with the allele frequencies.  That is: $p_{11}=p_1^2$, $p_{12}=2p_1(1-p_1)$ and $p_{22}=(1-p_1)^2$.  In your basic texts, they typicaly use $p$ and $q$ for the frequencies of allele 1 and 2, and present these *Hardy-Weinberg proportions* as: $p^2$,$2pq$, and $q^2$.  

Another way to think of the Hardy-Weinberg proportions is in the following way.  If the state of an allele ($A_1$ vs $A_2$) is *independent* within a genotype, then the probability of a particular genotype state (such as $A_1 A_1$) will be determined by taking the product of the alleles within it (so $p_{11}=p_1 p_1$ or $p_1^2$). 

Let's add to the plot lines that represent Hardy-Weinberg proportions:

```{r}
ggplot(gTidy)+
  geom_point(aes(x=p1,y=Genotype.Proportion,color=variable,shape=variable))+
  stat_function(fun=function(p) p^2, geom="line", colour="red",size=2.5) +
  stat_function(fun=function(p) 2*p*(1-p), geom="line", colour="green",size=2.5) +
  stat_function(fun=function(p) (1-p)^2, geom="line", colour="blue",size=2.5)
```

On average, the data follow the classic theoretical expectations fairly well.  It is pretty remarkable that such a simple theory has some bearing on reality! 

By eye, we can see that the fit isn't perfect though.  There is a systematic deficiency of heterozygotes and excess of homozygotes.  Why?  

Let's look at this more closely and more formally... 

### Testing Hardy Weinberg

Pearson's $\chi^2$-test is a basic statistical test that can be used to see if count data $o_i$ conform to a particular expectation.  It is based on the $X^2$-test statistic:
$$ X^2=\sum_i \frac{(o_i-e_i)^2}{e_i}$$
which follows a $\chi^2$ distribution under the null hypothesis that the data are generated from a multinomial distribution with the expected counts given by $e_i$. 

Here we compute the test statistic and obtain its associated p-value (using the `pchisq` function).  We keep in mind that there is 1 degree of freedom (because we have 3 observations per SNP, but then they have to sum to a single total sample size, and we have to use the data once to get the estimated allele frequency, which reduces us down to 1 degree of freedom).

```{r}
g <- mutate(g, X2 = (nA1A1-nObs*p1^2)^2 /(nObs*p1^2) + 
            (nA1A2-nObs*2*p1*p2)^2 / (nObs*2*p1*p2) + 
            (nA2A2-nObs*p2^2)^2 / (nObs*p2^2))
g <- mutate(g,pval = 1-pchisq(X2,1))
```


### The problem of multiple testing

Let's look at the top few p-values:

```{r}
head(g$pval)
```

How should we interpret these?  A p-value gives us the frequency at which the observed departure from expectations (or a more extreme departure) would occur if the null hypothesis is true.  As an agreed upon standard (of the frequentist paradigm for statistical hypothesis testing), if the observation is relatively rare under the null (e.g. p-value < 5%), we reject the null hypothesis, and we would infer that the given SNP departs from Hardy-Weinberg expectations.  This is problematic here though.   The problem is that we are testing many, many SNPs (Use `dim(g)` to remind yourself how many rows/SNPs are in the dataset).  Even if the null is universally true, 5% of our SNPs would be expected to be rejected using the standard frequentist paradigm.  This is called the multiple testing problem.  As an example, if we have 20,000 SNPs, that all obey the null hypothesis, we would on average naively reject the null for ~1000 SNPs based on the p-values < 0.05.   

We clearly need some methods to deal with the "multiple testing problem".  Two frameworks are the Bonferroni approach and false-discovery-rate (FDR) approaches.  We will not say more about these here.  Instead, we will do two simple checks to see though if our data are globally consistent with the null.  

First, let's see how many tests have p-values less than 0.05.  Is it much larger than the number we'd expect on average given the total number of SNPs and a 5% rate of rejection under the null?  

```{r}
sum(g$pval < 0.05, na.rm = TRUE)
```

Wow - we see many more.  This is our first sign that though by eye these data show qualitiative similarities to HW, statistically they are not fitting Hardy-Weinberg well enough. 

Let's look at this another way.  A classic result from Fisher is that under the null hypothesis the p-values of a well-designed test should be distributed uniformly between 0 and 1.  What do we see here? 

```{r}
qplot(pval, data = g)
```

The data show an enrichment for small p-values relative to a uniform distribution.  Notice how the whole distribution is shifted towards small values - The data appear to systematically depart from Hardy-Weinberg.  

### Plotting expected vs observed heterozygosity

To understand this more clearly, let's make a quick plot of the expected vs observed heterozygosity (the proportion of heterozygotes):
```{r}
qplot(2*p1*(1-p1), p12, data = g) + geom_abline(intercept = 0,
                                                slope=1,
                                                color="red",
                                                size=1.5)
```

Most of the points fall below the y=x line.  That is, we see a systematic deficiency of heterozygotes (and this implies a concordant excess of homozygotes).  This general pattern is contributing to the departure from HW seen in the $X^2$ statistics. 

### Discussion: Population subdivision and departures from Hardy-Weinberg expectations

We might wonder why the departure from Hardy-Weinberg proportional is directional, in that, on average, we are seeing a deficiency of heterozygotes (and excess of homozygotes).  One enlightening way to understand this is by thinking about what Sewall Wright (a former eminent University of Chicago professor) called "the correlation of uniting gametes".  To produce an $A_1 A_1$ individual we need an $A_1$-bearing sperm and an $A_1$-bearing egg to unite.  If these events were independent of each other, we would expect $A_1 A_1$ individuals at the rate predicted by multiplying probabilites, that is, $p_1^2$ (an idea we introduced above).  However, what if uniting gametes are positively correlated, in that an $A_1$-bearing sperm is more likely to join with an $A_1$-bearing egg?  In this case we will have more $A_1 A_1$ individuals than predicted by $p_1^2$, and conversely fewer $A_1 A_2$ individuals than predicted by $2 p_1 p_2$.  If our population is structured somehow such that $A_1$ sperm are more likely to meet with $A_1$ eggs, then we will have such a positive correlation of uniting gametes, and the resulting excess of homozygotes and deficiency of heterozygotes. 

Given the HGDP data is from 52 sub-populations from around the globe, and alleles have some some probability of clustering within populations, a good working hypothesis for the deficiency of heterozygotes in this dataset is the presence of some population structure.  

While statistically significant, the population structure appears to be subtle in absolute terms --- based on our plots, we have seen the genotype proportions are not wildly off from HW proportions. 

**Question:** As an exercise, compute the average deficiency of heterozygotes relative to the expectated proportion.  This is the average of $$ \frac{2 p_1 (1-p_1)-p_{12}}{2p_1(1-p_1)}$$  What is this number for this data-set?  A common "rule-of-thumb" for this deficiency in a global sample of humans is approximately 10\%.  Do you find this to be true from the data? 

```{r echo=FALSE}
mean(((2 * g$p1 * (1-g$p1)) - g$p12) / (2 * g$p1 * (1-g$p1)))
g <- mutate(g, F = (2*p1*(1-p1)-p12) / (2*p1*(1-p1)))
mean(g$F)
```
A ~10% difference between expected and observed seems pretty remarkable given these samples are taken from across the globe.  It is a reminder that human populations are not very deeply structured.  Most of the alleles in the sample are globally widespread and not sufficiently geographically clustered to generate correlations among the uniting alleles.  This is because all humans populations derived from an ancestral population in Africa around 100-150 thousand years ago, which is relatively small amount of time for variation across populations to accumulate. 

### Finding specific loci that show large departure from Hardy-Weinberg proportions

Now, let's ask if we can find any loci that are wild departures from HW proportions.  These might be loci that have erroneous genotypes, or loci that cluster geographically in dramatic ways (such that they have few heterozygotes relative to expectations).

To find these loci, we'll compute the same relative deficiency you computed above, but let's look at it per SNP.  This number is referred to as $F$ by Sewall Wright and has connections directly to correlation coefficients (advanced exercise: Try to work this out!).  If we assume there is no inbreeding within populations, this number is an estimator of $F_{ST}$ (a quantity that appears often in population genetics).

Let's add this value to our dataframe and plot how it's value changes across the chromosome from one end to another:

```{r}
g <- mutate(g, F = (2*p1*(1-p1)-p12) / (2*p1*(1-p1)))
plot(g$F, xlab = "SNP number")
```

There are a few interesting SNPS that show either a very high or low $F$ value. 

Now, here's a trick.  When a high or low $F$ value is due to genotyping error, it likely only effects a single SNP.  However, when there is some population genetic force acting on a region of the genome, it likely effects multiple SNPs in the region.  So let's try to take a local average in a sliding window of SNPs across the genome, computing an average $F$ over every 5 consecutive SNPs (in real data analysis we might use 100kb or 0.1cM windows). 

The `stats::filter` command below calls the `filter` function from the `stats` library.  The code instructs the function to take 5 values centered on a focal SNP, weighting them each by 1/5 and then taking the sum.  In this way it produces a local average in a sliding window of 5 SNPs.  Let's define the `movingavg` function and then make a plot of its values:
```{r}
movingavg <- function(x, n=5){stats::filter(x, rep(1/n,n), sides = 2)}
avgF <- movingavg(g$F)
plot(avgF, xlab="SNP number")
```

Wow --- there appears to be one large spike where the average $F$ is approximately 60% in the dataset!

Let's extract the SNP id for the largest value, and look at the dataframe:
```{r}
outlier=which (movingavg(g$F) == max(movingavg(g$F),na.rm=TRUE))
g[outlier,]
```

**Question:** Which SNP is returned?  By inserting the rs id into the UCSC genome browser (https://genome.ucsc.edu/), and following the links, find out what gene this SNP resides near.  The gene names should start with "SLC.."  What gene is it?

**Question:** Carry out a literature search on this gene using the term "positive selection" and see what you find. It's thought the high $F$ value observed here is because natural selection led to a geographic clustering of alleles in this gene region.  Discuss with your partners why this might or might not make sense.   


### Discussion: The outlier reigon

The region you've found is one of the most differentiated betwen human populations that is known.  Notice in your literature search, how it is known to affect skin pigmentation and is thought to contribute to differences in skin pigmentation that are seen between human populations.  Finding strong population structure for alleles that affect external morhological phenotypes is not uncommon when looking at other chromosomes.  Some of the most differentiated genes that exist in humans are those that involve morphological phenotypes - such as skin pigmentation, hair color/thickness, and eye color (the genes OCA2/HERC2, SCL45A2, KITLG, EDAR all come to mind).  Many of these are thought to have arisen due to direct or indirect effects of adaptation to local selective pressures (e.g. adaptation to varing levels of UV exposure, local pathogens, local diets, local mating preferences), though in most cases we still do not yet have a fully convincing understanding of their evolutionary histories.  Regardless of the reasons, it is notable that many of the features that humans see externally in each other (i.e. the morphological differences) are controlled by genes that are outliers in the genome.  At most variant SNPs, the patterns of variation are much closer to those of a single random mating populations than they are at variant sites like EDAR.  Put another way, a genomic perspective shows us many of the differences people see in each other are in a sense, just skin-deep. 

### Wrap-up

Modern population genetics has a lot of additional tools on its workbench, but here using relatively simple and classical ideas combined with genomic-scale data, we have been able to observe and interpret some major features of human genetic diversity.  We have also revisited some basic concepts of probablity and statistics such as indepedence vs correlation, the $\chi^2$ test, and the problems of multiple testing. One remarkable thing we saw is that a very simple mathematical model based on assuming independence of alleles and genotypes can predict genotype proportions within ~10% of the true values.  This gives us a hint of how simple mathematical models may be useful even in the face of biological complexity. Finally, we have gained more familiarity with R. We didn't discuss how genotyping errors might create Hardy-Weinberg departures, but if we were doing additional analyses, we could use Hardy-Weinberg departures to filter them from our data. It's common practice to do so, but with a Bonferonni correction and using data from within populations to do the filtering. 


### Follow-up activities

In the `data` folder, we are including data files that you can explore to gain more experience.  These include global data for other chromosomes (`H938_chr*.geno`) and the same data but limited to European populations (`H938_Euro_chr*.geno`).  Here are a few suggested follow-up activities.  It may be wise to split the activities across class members and reconvene after carrying them out.  

**Follow-up actvity**: Look at a chromosome from the European-restricted data - is the global deficiency in heterozygosity as strong as it was on the global scale?  Before you begin, what would you expect to see?  

**Follow-up activity** Using the European data, do you find any regions of the genome that are outliers for $F$ on chromosome 2?  Using genome browsers and/or literature searches, can you find what is the likely locus under selection for that region? 

**Follow-up activity**: Using the global data or the European data, analyze other chromosomes -- do you find other loci that show high $F$ values?


### References

Li, Jun Z, Devin M Absher, Hua Tang, Audrey M Southwick, Amanda M Casto, Sohini Ramachandran, Howard M Cann, et al. 2008. “Worldwide Human Relationships Inferred from Genome-Wide Patterns of Variation.” Science 319 (5866): 1100–1104.

Pickrell, Joseph K, Graham Coop, John Novembre, Sridhar Kudaravalli, Jun Z Li, Devin Absher, Balaji S Srinivasan, et al. 2009. “Signals of Recent Positive Selection in a Worldwide Sample of Human Populations.” Genome Research 19 (5): 826–37.